Connect Aiven for Apache Kafka with Klaw connection documentation (#1596)
aiven/devportal: docs/products/kafka/howto/kafka-klaw.rst

This article provides you with the steps to connect Aiven for Apache Kafka® service with Klaw. 

Prerequisites
-------------
To connect Aiven for Apache Kafka® and Klaw, you need to have the following setup: 

* A running Aiven for Apache Kafka® service. See :doc:`Getting started with Aiven for Apache Kafka </docs/products/kafka/getting-started>` for more information. 
* A running Klaw cluster. See `Run Klaw from the source <https://www.klaw-project.io/docs/run-source>`_ for more information.
* Configured :doc:`Java keystore and truststore containing the service SSL certificates </docs/products/kafka/howto/keystore-truststore>`. 

Connect Aiven for Apache Kafka® to Klaw
---------------------------------------

Follow the below steps to configure and connect Aiven for Apache Kafka® with Klaw: 

1. Log in to the **Klaw web interface**. 
2. Click **Environments**, and then click **Clusters**. 
3. On the **Clusters** page, click **Add Cluster** to add an Aiven for Apache Kafka® cluster. 
4. Enter the following details:
   
   -  **Cluster Type:** Select Kafka from the drop-down list
   - **Cluster Name:** Provide a name for the cluster

>    -  **Protocol:** Select the protocol for your cluster

Author: Fixed this.

   
   .. note:: 

>       Based on the protocol selected, you need to perform some :ref:`additional configurations <additional-configs>` to enable connection between Aiven for Apache Kafka® and Klaw clusters. 

Author: The advanced parameter for Apache Kafka is quite an exhaustive list. It makes sense to have them listed and defined in a separate topic (like how it is now). You can mention what advanced parameter the user needs to configure within the step and cross-link the relevant config/parameter in the advanced parameter topic for more details. 
In the case of Klaw, the content under additional configurations is necessary to ensure the connection between Apache Kafka and Klaw is successful. Currently, it is only the secret key config and authentication protocol. Therefore, I added them here to avoid users jumping between two different docs sites. (if that makes sense).

   -  **Kafka Flavor:** Select Aiven for Apache Kafka® as the flavor
   -  **Project Name:** Select the project name defined in the Aiven Console
   -  **Bootstrap server:** Enter the Service URI for your Apache Kafka service. You can find the service URI in the **Connection information** page of your service in Aiven Console. 
   - **Service Name:** Enter the name of the service as defined in the Aiven Console for your Apache Kafka service. 
5. Click **Save**.

> 6. Next, add the cluster to the preferred environment. Select **Evinorments**, and then select **Environments** from the drop-down menu. 

Author: I fixed the typo :). 
Yes, that is how it is on the UI.

7. Click **Add Environment** and enter the following details: 
   
   -  **Environment Name:** Select environment from the drop-down list.
   -  **Select Cluster:** Select the cluster you added from the drop-down list. The bootstrap servers and protocol details are automatically populated. 
   -  **Default Partitions:** Enter the number of partitions based on your requirements. The default value is set to 2.
   -  **Maximum Partitions:** Enter the maximum number of partitions. The default value is set to 2.
   -  **Default Replication factor:** Enter the required replication factor. The default value is set to 1.
   -  **Max Replication factor:** Enter the required max replication factor. The default value is set to 1.
   -  **Topic prefix (optional):** Enter a topic prefix.

>    -  **Tenant:** The value is set to default. 

Author: I added more information as a note. 
Klaw is multi-tenant. Each tenant can manage their topics with their own teams in isolation. Every tenant can have their own set of Kafka environments, and users of one tenant cannot view/access topics, ACLS, or any from other tenants. It provides isolation avoiding any security breach.

8. Click **Save**. 

If the connection based on the cluster and environment configuration are **successful**, the connection status displays a **blue thumbs-up icon** on the Environments page in the Klaw web interface.  If it is **unsuccessful**, a **red thumbs-down icon** is displayed. 

For unsuccessful connections, check the authentication protocol configurations. See the :ref:`Additional configurations <additional-configs>` for further details. After completing these configurations, check the connection status on the Environments page by clicking the thumbs-down icon to see if the setup was successful.

> .. _additional-configs:

Author: Done.

Additional configurations
-------------------------
`Klaw <https://www.klaw-project.io/>`_ uses the ``application.properties`` file to configure any application-related properties, such as secret key and authentication protocol configurations. Connecting Aiven for Apache Kafka® with Klaw requires you to perform the following additional configurations in the ``application.properties`` file.

Secret key configuration
~~~~~~~~~~~~~~~~~~~~~~~~

> Set the value of ``klaw.clusterapi.access.base64.secret`` with a secret key in the form of a Base64 encoded string in the ``application.properties`` file located in the following paths: 

Author: The user can basically set the value to any secret key of their choice but in the form of a Base64 encoded string.  So it depends on how the user/admin and the organization set up the base64 encoding.

1. Get the **Cluster ID** by clicking the copy icon on the **Clusters** page in the **Klaw web interface**.  
2. Next, open the application.properties file located in the **klaw/cluster-api/src/main/resources** directory. 
3. Configure the SSL properties to connect to Apache Kafka® clusters by editing the following lines:

   ::

      klawssl.kafkassl.keystore.location=client.keystore.p12
      klawssl.kafkassl.keystore.pwd=klaw1234
      klawssl.kafkassl.key.pwd=klaw1234
      klawssl.kafkassl.truststore.location=client.truststore.jks
      klawssl.kafkassl.truststore.pwd=klaw1234
      klawssl.kafkassl.keystore.type=pkcs12
      klawssl.kafkassl.truststore.type=JKS

   * For the lines starting with ``klawssl``, replace ``klawssl`` with the Klaw Cluster ID.
   * Replace ``client.keystore.p12`` with the path for the keystore and ``klaw1234`` with the password configured for the keystore file. 
   * Replace ``client.truststore.jks`` with the path for the truststore and ``klaw1234`` with the password configured for the truststore file. 
   * Save the ``application.properties`` file.

>    .. note:: To add multiple SSL configurations, copy and paste the above lines by prefixing them with the required cluster identification and relevant certificates.

Author: Added an example.

>    .. note:: To add multiple SSL configurations, copy and paste the above lines by prefixing them with the required cluster identification and relevant certificates.

Author: Added an example.

* A running Klaw cluster. See `Run Klaw from the source <https://www.klaw-project.io/docs/run-source>`_ for more information.
* Configured :doc:`Java keystore and truststore containing the service SSL certificates </docs/products/kafka/howto/keystore-truststore>`. 

Connect Aiven for Apache Kafka® to Klaw
---------------------------------------

Follow the below steps to configure and connect Aiven for Apache Kafka® with Klaw: 

1. Log in to the **Klaw web interface**. 
2. Click **Environments**, and then click **Clusters**. 
3. On the **Clusters** page, click **Add Cluster** to add an Aiven for Apache Kafka® cluster. 
4. Enter the following details:
   
   -  **Cluster Type:** Select Kafka from the drop-down list
   -  **Cluster Name:** Provide a name for the cluster
   -  **Protocol:** Select the protocol for your cluster
   
   .. note:: 
      Based on the protocol selected, you need to perform some :ref:`additional configurations <klaw-additional-configs>` to enable connection between Aiven for Apache Kafka® and Klaw clusters. 

   -  **Kafka Flavor:** Select Aiven for Apache Kafka® as the flavor
   -  **Project Name:** Select the project name defined in the Aiven Console

>    -  **Bootstrap server:** Enter the Service URI for your Apache Kafka service. You can find the service URI in the **Connection information** page of your service in Aiven Console 

Reviewer: Can we set a link to the Aiven console?

Author: added the link.

   -  **Service Name:** Enter the name of the service as defined in the Aiven Console for your Apache Kafka service
5. Click **Save**.
6. Next, add the cluster to the preferred environment. Select **Environments**, and then select **Environments** from the drop-down menu. 
7. Click **Add Environment** and enter the following details: 
   
   -  **Environment Name:** Select environment from the drop-down list
   -  **Select Cluster:** Select the cluster you added from the drop-down list. The bootstrap servers and protocol details are automatically populated 
   -  **Default Partitions:** Enter the number of partitions based on your requirements. The default value is set to 2
   -  **Maximum Partitions:** Enter the maximum number of partitions. The default value is set to 2
   -  **Default Replication factor:** Enter the required replication factor. The default value is set to 1
   -  **Max Replication factor:** Enter the required max replication factor. The default value is set to 1

>    -  **Topic prefix (optional):** Enter a topic prefix

Reviewer: What is this setting going to do? Is it going to prefix all created topics? Or is it a requirement for topic creation? Possibly giving a link to the klaw doc will help

Author: These are Kafka environment detail. Since in the previous steps, the user is adding a Kafka cluster to connect, this screen title says "Add Kafka Environment".  Modified step 7 to "Click **Add Environment** and enter the details to add Kafka environment". 

The Klaw is quite slim at the moment and is still a work in progress. However, I have cross-linked "Environment name" to the Klaw doc for more details. The rest of the parameters are known Kafka parameters. 
--

   -  **Tenant:** The value is set to default Tenant
   

>       .. note:: Klaw is multi-tenant. Each tenant can manage topics with their own teams in isolation. Every tenant can have their own set of Apache Kafka® environments, and users of one tenant cannot view/access topics, or ACLS from other tenants. It provides isolation avoiding any security breach. For this topic, I have used the default tenant configuration. For more information, see `Klaw documentation <https://www.klaw-project.io/docs/get-started-kafka#configure-the-cluster-to-sync>`_. 

Reviewer: ```suggestion
      .. note:: Klaw is multi-tenant by default. Each tenant manages topics with their own teams in isolation. Every tenant has its own set of Apache Kafka® environments, and users of one tenant cannot view/access topics, or ACLS from other tenants. It provides isolation avoiding any security breach. For more information, see `Klaw documentation <https://www.klaw-project.io/docs/get-started-kafka#configure-the-cluster-to-sync>`_. 
```

Author: Fixed this.

For unsuccessful connections, check the authentication protocol configurations. See the :ref:`Additional configurations <klaw-additional-configs>` section for further details. After completing these configurations, check the connection status on the Environments page by clicking the thumbs-down icon to see if the setup was successful.

.. _klaw-additional-configs:

Additional configurations
-------------------------
`Klaw <https://www.klaw-project.io/>`_ uses the ``application.properties`` file to configure any application-related properties, such as secret key and authentication protocol configurations. Connecting Aiven for Apache Kafka® with Klaw requires you to perform the following additional configurations in the ``application.properties`` file.

Secret key configuration
~~~~~~~~~~~~~~~~~~~~~~~~

Set the value of ``klaw.clusterapi.access.base64.secret`` with a secret key in the form of a Base64 encoded string in the ``application.properties`` file located in the following paths: 

* `klaw/cluter-api/src/main/resources`
* `klaw/core/src/main/resources`

Configure authentication protocol
""""""""""""""""""""""""""""""""""

> Klaw supports the following authentication protocols to connect to Apache Kafka® clusters:

Reviewer: I would say this is general to Klaw and not for Aiven for Apache Kafka and Klaw.
I would just set the two separate options as

* Connect Aiven for Apache Kafka® with Klaw via SSL protocol
* Connect Aiven for Apache Kafka® with Klaw via SASL protocol

Author: I would still keep this as part of the "Configure authentication protocol", the reason being this is a part of the "Additional configurations" for SSL in Klaw (and to give more context without jumping to other topics). 
If the protocol configurations are not done correctly, the connection between Klaw and Apache Kafka might not be successful.

1. Get the **Cluster ID** by clicking the copy icon on the **Clusters** page in the **Klaw web interface**.  
2. Next, open the application.properties file located in the **klaw/cluster-api/src/main/resources** directory. 
3. Configure the SSL properties to connect to Apache Kafka® clusters by editing the following lines:

   ::

      klawssl.kafkassl.keystore.location=client.keystore.p12
      klawssl.kafkassl.keystore.pwd=klaw1234
      klawssl.kafkassl.key.pwd=klaw1234
      klawssl.kafkassl.truststore.location=client.truststore.jks
      klawssl.kafkassl.truststore.pwd=klaw1234
      klawssl.kafkassl.keystore.type=pkcs12
      klawssl.kafkassl.truststore.type=JKS

   * For the lines starting with ``klawssl``, replace ``klawssl`` with the Klaw Cluster ID.
   * Replace ``client.keystore.p12`` with the path for the keystore and ``klaw1234`` with the password configured for the keystore file. 
   * Replace ``client.truststore.jks`` with the path for the truststore and ``klaw1234`` with the password configured for the truststore file. 
   * Save the ``application.properties`` file.

>    The following is an example of an ``application.properties`` file configured with Klaw Cluster ID (demo_cluster), keystore, and truststore paths and passwords. 

Reviewer: ```suggestion
   The following is an example of an ``application.properties`` file configured with Klaw Cluster ID (``demo_cluster``), keystore, and truststore paths and passwords. 
```

Reviewer: This change removes the need of defining ``demo_cluster`` in our dictionary

Author: Remove this.
